### DISCUSSION TITLE
Is as possible to catch an evaluation error (evaluate or evaluateLater) in a Directive?

### DISCUSSION THREAD

**Discussion by leobm** [alpinejs/alpine#2784] [1. Help] [answered]
Hi, 

I want evaluate an expression in a  directive. But when it failed e.g. with an Alpine Expression Error, I can't try/catch the 
the error. 

any solution for it?

thanks

**Comment by leobm** [comment, marked as answer]
I think my problem is this line here: https://github.com/alpinejs/alpine/blob/5cb40460d744c2a840f21d937e71e202471f7e21/packages/alpinejs/src/utils/error.js#L14

I might be able to catch (prevents the firing of the default event handle) the error via a global error handler, but i can't distinguish the error. 
```javascript
 window.onerror = function(message, source, lineno, colno, error) {
     console.log(message, error, source, lineno, colno);
     let suppressErrorAlert = true;
     return suppressErrorAlert;
};
```
Inside the error handler it would be just a normal Script error.

**Edit: 27.03.2022 12:44**

ok, what i just see that in this code line 
https://github.com/alpinejs/alpine/blob/5cb40460d744c2a840f21d937e71e202471f7e21/packages/alpinejs/src/utils/error.js#L10
actually to the error the element and the expression is added. Only I don't see these values in the error object which I get passed in the onerror handler.

not even if i catch the error like this:

``` javascript
window.addEventListener('error', (event) => {
       console.log(event);
       event.preventDefault();
});
```

**Reply by SimoTod** [reply]
Which kind of error do you want to catch exactly? A variable that doesn't exist? This errors are thrown because a developer is supposed to fix them before going live. Do you have a concrete example?

**Reply by leobm** [reply]
my specific example was this.
https://codepen.io/leobm/pen/dyJNOdB

i want to evaluate the expression, if that fails like in this case the expression itself should be the value.

i want to be able to write my directive on the element level once like this.

```html
  <a href="http://www.google.de" x-confirm="Do you want to open the link?">Google</a>
```

but also like this:
```html
  <a href="http://www.google.de" x-confirm="'Do you want to open the link?'">Google</a>
```

or like this
```html
<div x-data="{message: 'Do you want to open the link?'}">
  <a href="http://www.google.de" x-confirm="message">Google</a>
</div>
```

Ok, the example is not great. But I think it's a good idea if I could catch the evaluation error and then read out the 'expression" or the 'el' for example. There are certainly other reasons for it.

**Reply by leobm** [reply]
did you also see my hint about Object.assign above? if I had access to the 'expression' and 'el' value, then I could explicitly differentiate the error with other errors that have occurred. 

thanks

**Reply by SimoTod** [reply]
so... 

the 2 properties are available on event.error
```javascript
window.addEventListener('error', (event) => {
     console.log(event.error.el);
     console.log(event.error.expression);
     event.preventDefault();
});
 ```
 but that code is really complicated and it's never a good idea to mess with the error handler so I have 2 answers.
 
1) The answer you don't want to hear: basically, you shouldn't do it. The string should be valid javascript so wrapped in quotes. The whole point of the Alpine error handling is that if you forgot the quote, you'll get the error in the console and you can easily fix it. Skipping those quotes will save you a few seconds but the price is that you have to write code to handle both cases which can potentially contain bugs.

2) The answer you want to hear: Move the try-catch inside the expression itself. Your goal is to make sure that expression after eval returns a string, that's it. Your directive will be something like https://codepen.io/SimoTod/pen/eYyWYjb

**Reply by leobm** [reply]
Hello Simone, 
ok, strange, I must have done something wrong when I tried to read out the event values.
Maybe you're right, I'll do it the way you described in point 1.

Thanks to you

Greetings Felix
